### PR TITLE
Add template helpers to the op_reg facility.

### PR DESCRIPTION
**Pull request by stellaraccident**
* This support was ported and generalized from sharktank, where the technique originated.
* Support ops defined in terms of either a str-format or jinja2 based ASM template.
* Helpers for inlining and calling.
* Example ops used for testing.
* Adds dependency on Jinja2.

### CODE REVIEW

File: shark_turbine/runtime/op_reg/impl_helper.py [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+        *,
+        suffix: str = ".mlir",
+    ):
+        try:
+            from jinja2 import Environment, PackageLoader, select_autoescape
+        except ModuleNotFoundError as e:
+            raise ModuleNotFoundError(
+                "Cannot use JinjaTemplateLoader if jinja2 is not installed"
+            ) from e
+        self.env = Environment(loader=PackageLoader(package_name, package_path))
```
Comment (raikonenfnu) [inline shark_turbine/runtime/op_reg/impl_helper.py:160]: This is nice, so we can just specify the templates/env from the caller. :)

File: shark_turbine/ops/_str_format_test_ops.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+        result_desc.specialize_all_dims()
+
+    def generate(self, ksel: KernelSelection, kb: KernelBuilder):
+        result_type = kb.arg_bindings[0].type  # type: ignore
+        rtt = RankedTensorType(result_type)
+        function_name = (
+            f"turbine_test_add_strformat_{rtt.rank}d_{str(rtt.element_type)}"
+        )
+        func_op = _templates.inline_template_function(
+            kb,
```
Comment (raikonenfnu) [inline shark_turbine/ops/_str_format_test_ops.py:46]: Just curious, any special reasoning behind adding support for string-based templates?

Comment (stellaraccident) [reply shark_turbine/ops/_str_format_test_ops.py:46]: Not a huge one: it let me test everything without an external dep. I had used str format for all of the early work. Jinja is a better experience but just felt better having a path that was dep free.

File: tests/runtime/device_test.py [hunk 3]
```diff
@@ -151,20 +151,28 @@ def testFromTorchDevice(self):
         print(device.dump_device_info())
 
     def testJit(self):
-        from shark_turbine.ops import iree as iree_ops
+        from shark_turbine.ops import _str_format_test_ops as test_ops
```
Comment (raikonenfnu) [inline tests/runtime/device_test.py:154]: NIT: Can do later or set as TODO. Add test for jinja-based formatting and check output MLIR string from export perhaps. :)

File: shark_turbine/ops/_str_format_test_ops.py (hunk 2 above)
Comment (raikonenfnu) [reply shark_turbine/ops/_str_format_test_ops.py:46]: makes sense :)

**Review by raikonenfnu — APPROVED** [review]

File: tests/runtime/device_test.py (hunk 3 above)
Comment (stellaraccident) [reply tests/runtime/device_test.py:154]: There is one below on the CPU. I debated having more of a lower level test but in this case, I think a functional test gets the job done.

**Pull request merged**